File: packages/sdk-android/src/main/kotlin/com/mobiledevmcp/connection/WebSocketClient.kt
/**
 * WebSocket client for Android
 */

package com.mobiledevmcp.connection

import android.os.Build
import android.util.Log
import com.mobiledevmcp.MCPCommand
import kotlinx.coroutines.*
import okhttp3.*
import org.json.JSONObject
import java.util.concurrent.TimeUnit

class WebSocketClient(
    private val url: String,
    private val onCommand: suspend (MCPCommand) -> Any?
) {
    
    private val TAG = "MCP WebSocket"
    
    private val client = OkHttpClient.Builder()
        .readTimeout(0, TimeUnit.MILLISECONDS)
        .build()
    
    private var webSocket: WebSocket? = null
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    
    var isConnected = false
        private set
    
    fun connect() {
        val request = Request.Builder()
            .url(url)
            .build()
        
        webSocket = client.newWebSocket(request, object : WebSocketListener() {
            
            override fun onOpen(webSocket: WebSocket, response: Response) {
                Log.i(TAG, "Connected to server")
                isConnected = true
                sendHandshake()
            }
            
            override fun onMessage(webSocket: WebSocket, text: String) {
                handleMessage(text)
            }
            
            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                Log.i(TAG, "Disconnected: $code $reason")
                isConnected = false
                scheduleReconnect()
            }
            
            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                Log.e(TAG, "Connection failed: ${t.message}")
                isConnected = false
                scheduleReconnect()
            }
        })
    }
    
    fun disconnect() {
        webSocket?.close(1000, "Client disconnect")
        webSocket = null
        isConnected = false
        scope.cancel()
    }
    
    private fun sendHandshake() {
        val handshake = JSONObject().apply {
            put("type", "handshake")
            put("platform", "android")
            put("osVersion", Build.VERSION.SDK_INT)
            put("appName", "Unknown") // Would come from context
            put("appVersion", "0.0.0")
            put("capabilities", listOf("state", "network", "logs", "ui", "screenshot"))
        }
        send(handshake.toString())
    }
    
    private fun handleMessage(text: String) {
        try {
            val json = JSONObject(text)
            
            // Handle handshake acknowledgment
            if (json.optString("type") == "handshake_ack") {
                Log.d(TAG, "Handshake acknowledged")
                return
            }
            
            // Handle command
            val id = json.optString("id")
            val method = json.optString("method")
            
            if (id.isNotEmpty() && method.isNotEmpty()) {
                val paramsJson = json.optJSONObject("params") ?: JSONObject()
                val params = jsonObjectToMap(paramsJson)
                
                val command = MCPCommand(id, method, params)
                
                scope.launch {
                    try {
                        val result = onCommand(command)
                        sendResponse(id, result)
                    } catch (e: Exception) {
                        sendError(id, e.message ?: "Unknown error")
                    }
                }
            }
            
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse message: ${e.message}")
        }
    }
    
    private fun sendResponse(id: String, result: Any?) {
        val response = JSONObject().apply {
            put("jsonrpc", "2.0")
            put("id", id)
            put("result", when (result) {
                is Map<*, *> -> JSONObject(result as Map<String, Any?>)
                else -> result
            })
        }
        send(response.toString())
    }
    
    private fun sendError(id: String, message: String) {
        val response = JSONObject().apply {
            put("jsonrpc", "2.0")
            put("id", id)
            put("error", JSONObject().apply {
                put("code", -32000)
                put("message", message)
            })
        }
        send(response.toString())
    }
    
    private fun send(text: String) {
        webSocket?.send(text)
    }
    
    private fun scheduleReconnect() {
        scope.launch {
            delay(3000)
            if (!isConnected) {
                Log.i(TAG, "Attempting reconnection...")
                connect()
            }
        }
    }
    
    private fun jsonObjectToMap(json: JSONObject): Map<String, Any?> {
        val map = mutableMapOf<String, Any?>()
        json.keys().forEach { key ->
            map[key] = when (val value = json.get(key)) {
                is JSONObject -> jsonObjectToMap(value)
                JSONObject.NULL -> null
                else -> value
            }
        }
        return map
    }
}

File: packages/sdk-react-native/src/connection/WebSocketClient.ts
/**
 * WebSocket client for connecting to MCP server
 */

import type { MCPCommand, MCPResponse } from '../types';

interface WebSocketClientConfig {
  url: string;
  reconnectInterval: number;
  timeout: number;
  debug: boolean;
  onCommand: (command: MCPCommand) => Promise<unknown>;
}

export class WebSocketClient {
  private ws: WebSocket | null = null;
  private config: WebSocketClientConfig;
  private reconnectTimer: ReturnType<typeof setTimeout> | null = null;
  private connected = false;

  constructor(config: WebSocketClientConfig) {
    this.config = config;
  }

  connect(): void {
    if (this.ws?.readyState === WebSocket.OPEN) {
      return;
    }

    try {
      this.log('debug', `Connecting to ${this.config.url}...`);
      this.ws = new WebSocket(this.config.url);

      this.ws.onopen = () => {
        this.connected = true;
        this.log('info', 'Connected to MCP server');
        this.sendHandshake();
      };

      this.ws.onmessage = (event) => {
        this.handleMessage(event.data);
      };

      this.ws.onclose = (event) => {
        this.connected = false;
        this.log('warn', `Disconnected: ${event.code} ${event.reason}`);
        this.scheduleReconnect();
      };

      this.ws.onerror = (error) => {
        this.log('error', 'WebSocket error:', error);
      };

    } catch (error) {
      this.log('error', 'Failed to connect:', error);
      this.scheduleReconnect();
    }
  }

  disconnect(): void {
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }
    this.ws?.close();
    this.ws = null;
    this.connected = false;
  }

  isConnected(): boolean {
    return this.connected;
  }

  private sendHandshake(): void {
    const { Platform } = require('react-native');
    
    this.send({
      type: 'handshake',
      platform: 'react-native',
      osVersion: Platform.Version,
      appName: 'Unknown', // Would come from native module
      appVersion: '0.0.0',
      capabilities: [
        'state',
        'network',
        'logs',
        'ui',
        'screenshot',
      ],
    });
  }

  private async handleMessage(data: string): Promise<void> {
    try {
      const message = JSON.parse(data);

      // Handle handshake acknowledgment
      if (message.type === 'handshake_ack') {
        this.log('debug', `Handshake acknowledged, device ID: ${message.deviceId}`);
        return;
      }

      // Handle command from MCP server
      if (message.method && message.id) {
        try {
          const result = await this.config.onCommand(message as MCPCommand);
          this.sendResponse(message.id, result);
        } catch (error) {
          this.sendError(message.id, error instanceof Error ? error.message : 'Unknown error');
        }
      }

    } catch (error) {
      this.log('error', 'Failed to parse message:', error);
    }
  }

  private sendResponse(id: string, result: unknown): void {
    const response: MCPResponse = {
      jsonrpc: '2.0',
      id,
      result,
    };
    this.send(response);
  }

  private sendError(id: string, message: string): void {
    const response: MCPResponse = {
      jsonrpc: '2.0',
      id,
      error: {
        code: -32000,
        message,
      },
    };
    this.send(response);
  }

  private send(data: unknown): void {
    if (this.ws?.readyState === WebSocket.OPEN) {
      this.ws.send(JSON.stringify(data));
    }
  }

  private scheduleReconnect(): void {
    if (this.reconnectTimer) return;

    this.reconnectTimer = setTimeout(() => {
      this.reconnectTimer = null;
      this.connect();
    }, this.config.reconnectInterval);
  }

  private log(level: 'debug' | 'info' | 'warn' | 'error', message: string, ...args: unknown[]): void {
    if (!this.config.debug && level === 'debug') return;
    console[level](`[MCP WS] ${message}`, ...args);
  }
}
